lib/stages/basebandReadout.cpp

```diff
@@ -509,10 +510,32 @@ basebandDumpData::Status basebandReadout::extract_data(basebandDumpData data) {
                       frame_index, in_start, in_start + copy_len * _num_elements, stream_freq_idx,
                       out_frame_id, out_start, copy_len * _num_elements,
                       (in_start * _num_freq_per_stream + stream_freq_idx) * _num_elements);
-                memcpy(out_frame + (out_start * _num_elements),
-                       in_buf_data
-                           + (in_start * _num_freq_per_stream + stream_freq_idx) * _num_elements,
-                       copy_len * _num_elements);
+
+                if (_num_elements % 128 == 0) {
+                    // For other correlated data?  Assumes samples with all elements
+                    // are organized with time fastest, frequency slowest.
+                    memcpy(out_frame + (out_start * _num_elements),
+                           in_buf_data + (in_start * _num_freq_per_stream
+                                          + stream_freq_idx) * _num_elements,
+                           copy_len * _num_elements);
+                } else {
+                    // Direct from FPGA boards; each sample contains 16 channels for
+                    // different frequencies, so to get a time stream for a given
+                    // frequency we have to skip over the others.
+                    assert(_num_elements == 16);
+                    // Make pointers to 16-byte unsigned integers to be able to copy all
+                    // 16 elements in one go, leaving it to the compiler to optimize this.
+                    __int128* out_ptr = (__int128*)out_frame;
+                    __int128* in_ptr = (__int128*)in_buf_data;
+
+                    out_ptr += out_start;
+                    in_ptr += in_start;
+                    for (int i = 0; i < copy_len; i++) {
+                        *out_ptr = *in_ptr;
```

Comment:
There might be an advantage to using a `_mm_stream_load_si128` followed by an `_mm_stream_si128`.  Since we aren't going to be reading the output until later in another thread, using the NT hint for store reduces cache pressure.  And because the data read per frequency loop is likely too large for the cache, the prefetcher is likely hurting not helping.  The compiler is unlikely to add the NT hint on its own.

Reply:
The arrays are all allocated using an aligned memory allocator, so the alignment conditions for using those instructions should be met.